SeQuant/core/expr.cpp

```diff
@@ -9,6 +9,16 @@
 
 namespace sequant {
 
+ExprPtr ExprPtr::clone() const & {
+  if (!*this) return {};
+  return ExprPtr(as_shared_ptr()->clone());
+}
+
+ExprPtr ExprPtr::clone() && noexcept {
+  if (!*this) return {};
+  return std::move(*this);
```

Comment:
Return by move inhibits return-value optimization and therefore usually does more harm than good - see e.g.  https://stackoverflow.com/q/17473753

Reply:
the conventional wisdom has exceptions. This is one of them.

[try this](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIM6SuADJ4DJgAcj4ARpjE/gDspAAOqAqETgwe3r7%2ByanpAiFhkSwxcWaJdpgOGUIETMQEWT5%2BAVU1AnUNBEUR0bEJtvWNzTltwz2hfaUDFQCUtqhexMjsHOYAzKHI3lgA1CYbbk4KBMSYrIfYJhoAgje3p8ReDnu3B/FWd3s/e6EEeyYBw2ABE9gBaLiHL73b6/W4Qf6AuZ7ECAiBMFEmT57U7oEAoJYAw5uEkHMxmULpMR4ABemHQbz2AHdCAhAYcQeYzMDSUdAbyyXiCa5aNCPly4T8EVjQXssFQmF5aARoQ9fm8IGgGKc3uYAGx7VAEBCxFEMVCYVSrJIAtFMCDG03EAB0mI%2BFlxBHxhK8xKOZO5aCSAE89g4SNzBfzhSBReLsZL3vCIPczPqDUaTWa9harTa7einbE3ViccXXUDORCoRtPbG0H7o25yWYWKgAG6YcNEYhRklC70ihjoMV1iXq%2BGZ1BJWJMXucrUCXVpw0Vsueqtyitu8UNonN1vBsNMBRpYAMNiCfsBmNDuMjsee84EZYMPYAKhNeAUCfiSY1VcjVnYh50jUFUwNads2IDcBWrHcmHFRDgTBSE93vRt/T5FtuXbLtATPPALyvVUKUPWN43HF830/b9f3HRMHknaVw1oARMAgLF0w9PYaOId8EXo8ErkxP8APhNiOK48Nl2JHjsWfTBXwEzVhNEuZxJYpkdmk7iMwUnF%2BPfWN8M4r8EB/TTGP/bT3l0sIAH1mVQRyzJkqDDKUlT3wsn8tLuJi62Yu4kRYJhQg8z47MBDQIC4ayYUAwEuFQr0fXcpgNDmF0HMwZzXPcxLtP3JsB35bksrdNKbxw2LqvKltKMfNUpXSgksMPSquAauVarJJgeq3OrmtHVrk1YpgeWrUzO04rKcryrjxo1UrsMDCkqq3MF%2Bv5LaKPvKikt%2BNaus2sxep28jGsBC7hsHH0juY/8OAWWhOAAVl4PwOC0UhUE4UlLGsXElhWbtNh4UgCE0V6FgAaxADYNhdZG0fRjH9X0ThJG%2B2H/s4XgFBADRodhhY4FgJA0BYJI6FichKBpun6DiHZDGALh4g0Lg%2BDoAhYmJiAonxqJQgaENOChsXmGIEMAHkom0aoYe4XgadI%2BWGFoSXft4LAoi8YA3DEWhibV0gsHCoxxD1y28HOGou3Nv6rWqP01j%2B/5MHeu3aDwKJQLljwsHxs48BYKXXr4AxgAUAA1PBMGZeXZx%2BqH%2BEEEQxHYKQZEERQVHUO3dF5gwjBQaxrH0APicgBYZ0cZdCf%2BrtiGIPAsDrrjbB9lWMhcEdRj8XngimEoyj0FI0ibzJPBaKf8ln3oJ4GXn2lnroRnnnJ177joGC3yZin6OJ14mYe9FOboV9PiQFgUMHVnv7GOC%2B0gfr%2BgGOD2VQAA59Tgn1JINiHM9hcxdBoF0qUIC4EICQckGwEq8FVloOYCMkYowxtgtGWNfa4w/vjb%2BRMSZkz1hTGAiAQBEiSH6RmS5ab02IOEVgawACyoQADiAB1UBRhwH6kgS6MwvAGTwI7viXmmdhCiHEHnaRhc1D41LqQZkoEkhRzep9PGdtv7yz9LQgEqAqC/wAUAkB7N%2BEQKgTAjwjDWaIOQWQtBGDkaoxwdg1%2BBDP68GIbYUhqC4avxEYQ3RLdAnoNIG3NIzhJBAA%3D%3D)